### PR TITLE
database close should stop active listener

### PR DESCRIPTION
**Pull request by jayahariv**
* database close should stop the listener; feature implemented
* include a unit test validate same

### CODE REVIEW

File: Objective-C/CBLDatabase.mm [hunk 1]
```diff
@@ -360,7 +361,7 @@ - (BOOL) inBatch: (NSError**)outError usingBlock: (void (NS_NOESCAPE ^)())block
 #pragma mark - DATABASE MAINTENANCE
 
 - (BOOL) close: (NSError**)outError {
-    NSArray *activeReplicators, *activeLiveQueries = nil;
+    NSArray *activeReplicators, *activeLiveQueries, *activeListeners = nil;
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm:364]: Can you add EE ifdef?

File: Objective-C/Internal/CBLDatabase+Internal.h [hunk 2]
```diff
@@ -67,6 +68,10 @@ NS_ASSUME_NONNULL_BEGIN
 - (void) removeActiveLiveQuery: (CBLLiveQuery*)liveQuery;
 - (uint64_t) activeLiveQueryCount; // For testing only
 
+- (void) addActiveListener: (CBLURLEndpointListener*)listener API_AVAILABLE(macos(10.12));
```
Comment (pasin) [inline Objective-C/Internal/CBLDatabase+Internal.h:71]: Can you add EE ifdef?

File: Objective-C/CBLDatabase.mm [hunk 3]
```diff
@@ -1083,6 +1096,34 @@ - (uint64_t) activeReplicatorCount {
     }
 }
 
+#pragma mark - Listener
+#ifdef COUCHBASE_ENTERPRISE
+- (void) addActiveListener: (CBLURLEndpointListener*)listener API_AVAILABLE(macos(10.12)) {
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm:1101]: I think we need API_AVAILABLE for iOS 10 as well?

File: Objective-C/CBLDatabase.mm [hunk 4]
```diff
@@ -1083,6 +1096,34 @@ - (uint64_t) activeReplicatorCount {
...
+    CBL_LOCK(self) {
+        [self mustBeOpenAndNotClosing];
+        if (!_activeListeners)
+            _activeListeners = [NSMutableSet new];
+        
+        [_activeListeners addObject: listener];
+    }
+}
+
+- (void) removeActiveListener: (CBLURLEndpointListener*)listener API_AVAILABLE(macos(10.12)) {
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm:1111]: I think we need API_AVAILABLE for iOS 10 as well?

File: Objective-C/Internal/CBLDatabase+Internal.h (hunk 2 above)
Comment (pasin) [reply Objective-C/Internal/CBLDatabase+Internal.h:71]: I think we need API_AVAILABLE for iOS 10 as well?

File: Objective-C/Internal/CBLDatabase+Internal.h [hunk 5]
```diff
@@ -67,6 +68,10 @@ NS_ASSUME_NONNULL_BEGIN
...
+- (void) removeActiveListener: (CBLURLEndpointListener*)listener API_AVAILABLE(macos(10.12));
```
Comment (pasin) [inline Objective-C/Internal/CBLDatabase+Internal.h:72]: I think we need API_AVAILABLE for iOS 10 as well?

File: Objective-C/CBLDatabase.mm [hunk 6]
```diff
@@ -65,6 +65,7 @@ @implementation CBLDatabase {
     
     NSMutableSet<CBLReplicator*>* _activeReplicators;
     NSMutableSet<CBLLiveQuery*>* _activeLiveQueries;
+    NSMutableSet* _activeListeners;
```
[the remark this answers is not in the document]
Comment (jayahariv) [reply Objective-C/CBLDatabase.mm:68]: type would force the API check(ios 10, mac 12) for CBLDatabase implementation, thats why kept it without type.

[the remark this answers is not in the document]
Comment (pasin) [reply Objective-C/CBLDatabase.mm:69]: Can you add API_AVAILABLE(ios(10), macosx(10.12)) to the variable?

[the remark this answers is not in the document]
Comment (pasin) [reply Objective-C/CBLDatabase.mm:69]: Can you add API_AVAILABLE(ios(10), macosx(10.12)) to the variable? I haven't tried but I guessed you can do it. It's fine here if you can't do it.

**Pull request merged**